File: frontend/src/services/templates.ts
/**
 * Templates API Service
 * Handles all template-related API calls including browsing, downloading, and bookmarks
 */

import { apiGet, apiPost, apiPatch, apiDelete } from './api'
import type {
  AgentTemplateItem,
  AgentTemplateDetail,
  AgentTemplateWithUserData,
  TemplateQueryParams,
  TemplateCreateData,
  TemplateUpdateData,
  TemplateBookmarkStatus,
  TemplateDownloadResponse,
  TemplateStats,
  ApiAgentTemplateItem,
  ApiAgentTemplateDetail,
  TemplateCategory,
  TemplateDifficulty,
  TemplateStatus,
} from '@/types/models'

// ============================================================================
// Response Transformers (snake_case from API -> camelCase for frontend)
// ============================================================================

function transformTemplateItem(item: ApiAgentTemplateItem): AgentTemplateItem {
  return {
    id: item.id,
    name: item.name,
    description: item.description,
    category: item.category as TemplateCategory,
    difficulty: item.difficulty as TemplateDifficulty,
    tags: item.tags,
    authorName: item.author_name,
    status: item.status as TemplateStatus,
    featured: item.featured,
    downloadCount: item.download_count,
    model: item.model,
    hasTools: item.has_tools,
    hasSubAgents: item.has_sub_agents,
    thumbnailUrl: item.thumbnail_url,
    createdAt: item.created_at,
  }
}

function transformTemplateDetail(item: ApiAgentTemplateDetail): AgentTemplateDetail {
  return {
    ...transformTemplateItem(item),
    yamlContent: item.yaml_content,
    useCase: item.use_case,
    authorId: item.author_id,
    rejectionReason: item.rejection_reason,
    approvedBy: item.approved_by,
    approvedAt: item.approved_at,
    updatedAt: item.updated_at,
  }
}

function transformTemplateWithUserData(
  item: ApiAgentTemplateDetail & { is_bookmarked: boolean }
): AgentTemplateWithUserData {
  return {
    ...transformTemplateDetail(item),
    isBookmarked: item.is_bookmarked,
  }
}

// ============================================================================
// Request Transformers (camelCase from frontend -> snake_case for API)
// ============================================================================

function transformCreateData(data: TemplateCreateData): Record<string, unknown> {
  return {
    name: data.name,
    description: data.description,
    yaml_content: data.yamlContent,
    category: data.category,
    difficulty: data.difficulty,
    use_case: data.useCase,
    tags: data.tags,
    thumbnail_url: data.thumbnailUrl,
  }
}

function transformUpdateData(data: TemplateUpdateData): Record<string, unknown> {
  const result: Record<string, unknown> = {}
  if (data.name !== undefined) result.name = data.name
  if (data.description !== undefined) result.description = data.description
  if (data.yamlContent !== undefined) result.yaml_content = data.yamlContent
  if (data.category !== undefined) result.category = data.category
  if (data.difficulty !== undefined) result.difficulty = data.difficulty
  if (data.useCase !== undefined) result.use_case = data.useCase
  if (data.tags !== undefined) result.tags = data.tags
  if (data.thumbnailUrl !== undefined) result.thumbnail_url = data.thumbnailUrl
  return result
}

// ============================================================================
// Public Endpoints (Authenticated Users)
// ============================================================================

/**
 * Get approved templates with optional filters
 */
export async function getTemplates(
  params?: TemplateQueryParams
): Promise<AgentTemplateItem[]> {
  const apiParams: Record<string, unknown> = {}
  if (params) {
    if (params.skip !== undefined) apiParams.skip = params.skip
    if (params.limit !== undefined) apiParams.limit = params.limit
    if (params.search) apiParams.search = params.search
    if (params.category) apiParams.category = params.category
    if (params.difficulty) apiParams.difficulty = params.difficulty
    if (params.featured !== undefined) apiParams.featured = params.featured
    if (params.hasTools !== undefined) apiParams.has_tools = params.hasTools
    if (params.hasSubAgents !== undefined) apiParams.has_sub_agents = params.hasSubAgents
  }

  const templates = await apiGet<ApiAgentTemplateItem[]>('/templates', apiParams)
  return templates.map(transformTemplateItem)
}

/**
 * Get featured templates
 */
export async function getFeaturedTemplates(limit = 6): Promise<AgentTemplateItem[]> {
  const templates = await apiGet<ApiAgentTemplateItem[]>('/templates/featured', { limit })
  return templates.map(transformTemplateItem)
}

/**
 * Get a single template by ID with user bookmark status
 */
export async function getTemplate(id: string): Promise<AgentTemplateWithUserData> {
  const template = await apiGet<ApiAgentTemplateDetail & { is_bookmarked: boolean }>(
    `/templates/${id}`
  )
  return transformTemplateWithUserData(template)
}

/**
 * Download template YAML (increments download count)
 */
export async function downloadTemplate(id: string): Promise<TemplateDownloadResponse> {
  const response = await apiGet<{
    filename: string
    content: string
    content_type: string
  }>(`/templates/${id}/download`)
  return {
    filename: response.filename,
    content: response.content,
    contentType: response.content_type,
  }
}

/**
 * Get raw YAML content (for clipboard, doesn't increment download count)
 */
export async function getTemplateYaml(id: string): Promise<string> {
  const response = await fetch(`/api/v1/templates/${id}/yaml`, {
    headers: {
      Authorization: `Bearer ${localStorage.getItem('access_token')}`,
      'X-Tenant-ID': localStorage.getItem('tenant_id') || '',
    },
  })
  if (!response.ok) throw new Error('Failed to get template YAML')
  return response.text()
}

/**
 * Toggle bookmark for a template
 */
export async function toggleTemplateBookmark(id: string): Promise<TemplateBookmarkStatus> {
  const response = await apiPost<{
    is_bookmarked: boolean
    bookmarked_at: string | null
  }>(`/templates/${id}/bookmark`, {})
  return {
    isBookmarked: response.is_bookmarked,
    bookmarkedAt: response.bookmarked_at,
  }
}

/**
 * Get user's bookmarked templates
 */
export async function getBookmarkedTemplates(): Promise<AgentTemplateItem[]> {
  const templates = await apiGet<ApiAgentTemplateItem[]>('/templates/bookmarks')
  return templates.map(transformTemplateItem)
}

/**
 * Get user's submitted templates (all statuses)
 */
export async function getMySubmissions(): Promise<AgentTemplateDetail[]> {
  const templates = await apiGet<ApiAgentTemplateDetail[]>('/templates/my-submissions')
  return templates.map(transformTemplateDetail)
}

// ============================================================================
// Instructor Endpoints
// ============================================================================

/**
 * Submit a new template for review
 */
export async function submitTemplate(data: TemplateCreateData): Promise<AgentTemplateDetail> {
  const template = await apiPost<ApiAgentTemplateDetail>(
    '/templates',
    transformCreateData(data)
  )
  return transformTemplateDetail(template)
}

/**
 * Update own template (only draft or rejected)
 */
export async function updateMyTemplate(
  id: string,
  data: TemplateUpdateData
): Promise<AgentTemplateDetail> {
  const template = await apiPatch<ApiAgentTemplateDetail>(
    `/templates/${id}`,
    transformUpdateData(data)
  )
  return transformTemplateDetail(template)
}

/**
 * Delete own draft template
 */
export async function deleteMyTemplate(id: string): Promise<void> {
  await apiDelete(`/templates/${id}`)
}

// ============================================================================
// Admin Endpoints
// ============================================================================

/**
 * Get templates pending review
 */
export async function getPendingTemplates(): Promise<AgentTemplateDetail[]> {
  const templates = await apiGet<ApiAgentTemplateDetail[]>('/templates/admin/pending')
  return templates.map(transformTemplateDetail)
}

/**
 * Get all templates (including non-approved) - admin only
 */
export async function getAllTemplates(
  params?: TemplateQueryParams
): Promise<AgentTemplateDetail[]> {
  const apiParams: Record<string, unknown> = {}
  if (params) {
    if (params.skip !== undefined) apiParams.skip = params.skip
    if (params.limit !== undefined) apiParams.limit = params.limit
    if (params.search) apiParams.search = params.search
    if (params.category) apiParams.category = params.category
    if (params.status) apiParams.status = params.status
  }

  const templates = await apiGet<ApiAgentTemplateDetail[]>('/templates/admin/all', apiParams)
  return templates.map(transformTemplateDetail)
}

/**
 * Get template statistics
 */
export async function getTemplateStats(): Promise<TemplateStats> {
  const stats = await apiGet<{
    total_templates: number
    approved_templates: number
    pending_templates: number
    total_downloads: number
    templates_by_category: Record<string, number>
    templates_by_difficulty: Record<string, number>
  }>('/templates/stats')
  return {
    totalTemplates: stats.total_templates,
    approvedTemplates: stats.approved_templates,
    pendingTemplates: stats.pending_templates,
    totalDownloads: stats.total_downloads,
    templatesByCategory: stats.templates_by_category,
    templatesByDifficulty: stats.templates_by_difficulty,
  }
}

/**
 * Approve a template
 */
export async function approveTemplate(id: string): Promise<AgentTemplateDetail> {
  const template = await apiPost<ApiAgentTemplateDetail>(`/templates/${id}/approve`, {})
  return transformTemplateDetail(template)
}

/**
 * Reject a template with reason
 */
export async function rejectTemplate(id: string, reason: string): Promise<AgentTemplateDetail> {
  const template = await apiPost<ApiAgentTemplateDetail>(`/templates/${id}/reject`, { reason })
  return transformTemplateDetail(template)
}

/**
 * Set featured status for a template
 */
export async function setTemplateFeatured(
  id: string,
  featured: boolean
): Promise<AgentTemplateDetail> {
  const template = await apiPost<ApiAgentTemplateDetail>(`/templates/${id}/feature`, {
    featured,
  })
  return transformTemplateDetail(template)
}

/**
 * Update any template (admin)
 */
export async function adminUpdateTemplate(
  id: string,
  data: TemplateUpdateData
): Promise<AgentTemplateDetail> {
  const template = await apiPatch<ApiAgentTemplateDetail>(
    `/templates/admin/${id}`,
    transformUpdateData(data)
  )
  return transformTemplateDetail(template)
}

/**
 * Delete any template (admin)
 */
export async function adminDeleteTemplate(id: string): Promise<void> {
  await apiDelete(`/templates/admin/${id}`)
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Download YAML file to user's computer
 */
export function downloadYamlFile(content: string, filename: string): void {
  const blob = new Blob([content], { type: 'text/yaml' })
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  document.body.removeChild(a)
  URL.revokeObjectURL(url)
}

/**
 * Copy text to clipboard
 */
export async function copyToClipboard(text: string): Promise<boolean> {
  try {
    await navigator.clipboard.writeText(text)
    return true
  } catch {
    // Fallback for older browsers
    const textArea = document.createElement('textarea')
    textArea.value = text
    textArea.style.position = 'fixed'
    textArea.style.left = '-9999px'
    document.body.appendChild(textArea)
    textArea.select()
    try {
      document.execCommand('copy')
      return true
    } catch {
      return false
    } finally {
      document.body.removeChild(textArea)
    }
  }
}

/**
 * Get display label for category
 */
export const CATEGORY_LABELS: Record<TemplateCategory, string> = {
  hr: 'Human Resources',
  scheduling: 'Scheduling',
  faq: 'FAQ & Knowledge',
  'customer-service': 'Customer Service',
  'data-entry': 'Data Entry',
  workflow: 'Workflow Automation',
  healthcare: 'Healthcare',
  other: 'Other',
}

/**
 * Get display label for difficulty
 */
export const DIFFICULTY_LABELS: Record<TemplateDifficulty, string> = {
  beginner: 'Beginner',
  intermediate: 'Intermediate',
  advanced: 'Advanced',
}

/**
 * Get color for difficulty badge
 */
export const DIFFICULTY_COLORS: Record<TemplateDifficulty, string> = {
  beginner: 'bg-green-100 text-green-800',
  intermediate: 'bg-yellow-100 text-yellow-800',
  advanced: 'bg-red-100 text-red-800',
}

/**
 * Get color for status badge
 */
export const STATUS_COLORS: Record<TemplateStatus, string> = {
  draft: 'bg-gray-100 text-gray-800',
  pending_review: 'bg-blue-100 text-blue-800',
  approved: 'bg-green-100 text-green-800',
  rejected: 'bg-red-100 text-red-800',
}

File: src/api/routes/templates.py
"""Agent Template management API routes.

Templates are stored in the shared schema and accessible by all tenants.
Bookmarks are tenant-scoped (per-user within tenant).
"""

import logging
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.api.dependencies import (
    get_current_user,
    get_tenant_db_dependency,
    get_tenant_id,
    require_role,
)
from src.api.schemas.template import (
    AgentTemplateCreate,
    AgentTemplateListResponse,
    AgentTemplateResponse,
    AgentTemplateUpdate,
    AgentTemplateWithUserData,
    TemplateApproveRequest,
    TemplateBookmarkStatusResponse,
    TemplateDownloadResponse,
    TemplateFeatureRequest,
    TemplateFilters,
    TemplateRejectRequest,
    TemplateStatsResponse,
)
from src.core.constants import (
    TemplateCategory,
    TemplateDifficulty,
    TemplateStatus,
    UserRole,
)
from src.core.exceptions import NotFoundError, ValidationError
from src.core.tenancy import TenantContext
from src.db.models.user import User
from src.services.template_service import AgentTemplateService, TemplateBookmarkService

logger = logging.getLogger(__name__)

router = APIRouter()

# Role dependencies
require_super_admin = require_role(UserRole.SUPER_ADMIN)
require_instructor = require_role(UserRole.INSTRUCTOR)


# ============================================================================
# Public Endpoints (Authenticated Users)
# ============================================================================


@router.get("", response_model=list[AgentTemplateListResponse])
async def list_templates(
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    skip: Annotated[int, Query(ge=0)] = 0,
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
    search: Annotated[str | None, Query()] = None,
    category: Annotated[TemplateCategory | None, Query()] = None,
    difficulty: Annotated[TemplateDifficulty | None, Query()] = None,
    featured: Annotated[bool | None, Query()] = None,
    has_tools: Annotated[bool | None, Query()] = None,
    has_sub_agents: Annotated[bool | None, Query()] = None,
) -> list[AgentTemplateListResponse]:
    """
    List approved agent templates with pagination and filtering.

    Only returns approved templates for regular users.
    """
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    filters = TemplateFilters(
        search=search,
        category=category,
        difficulty=difficulty,
        featured=featured,
        has_tools=has_tools,
        has_sub_agents=has_sub_agents,
        status=TemplateStatus.APPROVED,  # Only approved for public listing
    )

    templates = await template_service.list_templates(
        skip=skip, limit=limit, filters=filters, include_pending=False
    )

    return [AgentTemplateListResponse.model_validate(t) for t in templates]


@router.get("/featured", response_model=list[AgentTemplateListResponse])
async def get_featured_templates(
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    limit: Annotated[int, Query(ge=1, le=12)] = 6,
) -> list[AgentTemplateListResponse]:
    """Get featured agent templates."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    templates = await template_service.get_featured_templates(limit=limit)

    return [AgentTemplateListResponse.model_validate(t) for t in templates]


@router.get("/my-submissions", response_model=list[AgentTemplateResponse])
async def get_my_submissions(
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> list[AgentTemplateResponse]:
    """Get templates submitted by the current user."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    templates = await template_service.get_user_templates(current_user.id)

    return [AgentTemplateResponse.model_validate(t) for t in templates]


@router.get("/bookmarks", response_model=list[AgentTemplateListResponse])
async def get_bookmarked_templates(
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> list[AgentTemplateListResponse]:
    """Get templates bookmarked by the current user."""
    TenantContext.set(tenant_id)

    bookmark_service = TemplateBookmarkService(db, tenant_id)
    template_service = AgentTemplateService(db)

    bookmarked_ids = await bookmark_service.get_bookmarked_template_ids(current_user.id)

    # Fetch each bookmarked template
    templates = []
    for template_id in bookmarked_ids:
        template = await template_service.get_template_by_id(template_id)
        if template and template.status == TemplateStatus.APPROVED.value:
            templates.append(template)

    return [AgentTemplateListResponse.model_validate(t) for t in templates]


@router.get("/stats", response_model=TemplateStatsResponse)
async def get_template_stats(
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> TemplateStatsResponse:
    """Get template statistics (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    stats = await template_service.get_template_stats()

    return TemplateStatsResponse(**stats)


@router.get("/{template_id}", response_model=AgentTemplateWithUserData)
async def get_template(
    template_id: str,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> dict:
    """Get a single template by ID with user bookmark status."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    template = await template_service.get_template_by_id(template_id)

    if not template:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Template with ID '{template_id}' not found",
        )

    # Only allow viewing approved templates (or own submissions)
    if template.status != TemplateStatus.APPROVED.value:
        if template.author_id != current_user.id and current_user.role != UserRole.SUPER_ADMIN.value:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Template with ID '{template_id}' not found",
            )

    # Get bookmark status
    bookmark_service = TemplateBookmarkService(db, tenant_id)
    is_bookmarked = await bookmark_service.is_bookmarked(current_user.id, template_id)

    # Convert to dict and add bookmark status
    template_data = AgentTemplateResponse.model_validate(template).model_dump()
    template_data["is_bookmarked"] = is_bookmarked

    return template_data


@router.get("/{template_id}/download", response_model=TemplateDownloadResponse)
async def download_template(
    template_id: str,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> TemplateDownloadResponse:
    """
    Download template YAML content.

    Increments the download counter.
    """
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    template = await template_service.get_template_by_id(template_id)

    if not template or template.status != TemplateStatus.APPROVED.value:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Template with ID '{template_id}' not found",
        )

    # Increment download count
    await template_service.increment_download_count(template_id)

    # Generate filename from template name
    safe_name = template.name.lower().replace(" ", "_").replace("-", "_")
    filename = f"{safe_name}.yaml"

    return TemplateDownloadResponse(
        filename=filename,
        content=template.yaml_content,
        content_type="text/yaml",
    )


@router.get("/{template_id}/yaml")
async def get_template_yaml(
    template_id: str,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> Response:
    """
    Get raw YAML content for clipboard copy.

    Does NOT increment download counter (use /download for that).
    """
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    template = await template_service.get_template_by_id(template_id)

    if not template or template.status != TemplateStatus.APPROVED.value:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Template with ID '{template_id}' not found",
        )

    return Response(
        content=template.yaml_content,
        media_type="text/yaml",
    )


@router.post("/{template_id}/bookmark", response_model=TemplateBookmarkStatusResponse)
async def toggle_template_bookmark(
    template_id: str,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(get_current_user)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> TemplateBookmarkStatusResponse:
    """Toggle bookmark status for a template."""
    TenantContext.set(tenant_id)

    # Verify template exists and is approved
    template_service = AgentTemplateService(db)
    template = await template_service.get_template_by_id(template_id)

    if not template or template.status != TemplateStatus.APPROVED.value:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Template with ID '{template_id}' not found",
        )

    bookmark_service = TemplateBookmarkService(db, tenant_id)
    bookmark = await bookmark_service.toggle_bookmark(current_user.id, template_id)

    return TemplateBookmarkStatusResponse(
        is_bookmarked=bookmark is not None,
        bookmarked_at=bookmark.created_at if bookmark else None,
    )


# ============================================================================
# Instructor Endpoints (Submit Templates)
# ============================================================================


@router.post("", response_model=AgentTemplateResponse, status_code=status.HTTP_201_CREATED)
async def submit_template(
    template_data: AgentTemplateCreate,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_instructor)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> AgentTemplateResponse:
    """
    Submit a new agent template for review.

    Requires instructor role or higher.
    Template will be in 'pending_review' status until approved by an admin.
    """
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)

    try:
        template = await template_service.create_template(
            template_data=template_data,
            author_id=current_user.id,
            author_name=current_user.full_name or current_user.email,
            status=TemplateStatus.PENDING_REVIEW,
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        ) from e

    logger.info(f"Template '{template.name}' submitted by {current_user.email}")

    return AgentTemplateResponse.model_validate(template)


@router.patch("/{template_id}", response_model=AgentTemplateResponse)
async def update_my_template(
    template_id: str,
    template_data: AgentTemplateUpdate,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_instructor)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> AgentTemplateResponse:
    """
    Update own template (only if draft or rejected).

    Instructors can only edit their own templates that haven't been approved yet.
    """
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    template = await template_service.get_template_by_id(template_id)

    if not template:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Template with ID '{template_id}' not found",
        )

    # Check ownership
    if template.author_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You can only edit your own templates",
        )

    # Can only edit draft or rejected templates
    if template.status not in [TemplateStatus.DRAFT.value, TemplateStatus.REJECTED.value]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Can only edit draft or rejected templates",
        )

    try:
        updated = await template_service.update_template(template_id, template_data)
        # Reset to pending review after edit
        updated.status = TemplateStatus.PENDING_REVIEW.value
        updated.rejection_reason = None
        await db.commit()
        await db.refresh(updated)
    except (NotFoundError, ValidationError) as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        ) from e

    return AgentTemplateResponse.model_validate(updated)


@router.delete("/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_my_template(
    template_id: str,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_instructor)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> None:
    """
    Delete own draft template.

    Instructors can only delete their own templates that are in draft status.
    """
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    template = await template_service.get_template_by_id(template_id)

    if not template:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Template with ID '{template_id}' not found",
        )

    # Check ownership
    if template.author_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You can only delete your own templates",
        )

    # Can only delete draft templates
    if template.status != TemplateStatus.DRAFT.value:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Can only delete draft templates",
        )

    await template_service.delete_template(template_id)


# ============================================================================
# Admin Endpoints (Manage All Templates)
# ============================================================================


@router.get("/admin/pending", response_model=list[AgentTemplateResponse])
async def get_pending_templates(
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> list[AgentTemplateResponse]:
    """Get all templates pending review (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    templates = await template_service.get_pending_templates()

    return [AgentTemplateResponse.model_validate(t) for t in templates]


@router.get("/admin/all", response_model=list[AgentTemplateResponse])
async def list_all_templates(
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    skip: Annotated[int, Query(ge=0)] = 0,
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
    status_filter: Annotated[TemplateStatus | None, Query(alias="status")] = None,
    search: Annotated[str | None, Query()] = None,
    category: Annotated[TemplateCategory | None, Query()] = None,
) -> list[AgentTemplateResponse]:
    """List all templates including non-approved (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)
    filters = TemplateFilters(
        search=search,
        category=category,
        status=status_filter,
    )

    templates = await template_service.list_templates(
        skip=skip, limit=limit, filters=filters, include_pending=True
    )

    return [AgentTemplateResponse.model_validate(t) for t in templates]


@router.post("/{template_id}/approve", response_model=AgentTemplateResponse)
async def approve_template(
    template_id: str,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    _: TemplateApproveRequest = TemplateApproveRequest(),
) -> AgentTemplateResponse:
    """Approve a pending template (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)

    try:
        template = await template_service.approve_template(template_id, current_user.id)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e

    logger.info(f"Template '{template.name}' approved by {current_user.email}")

    return AgentTemplateResponse.model_validate(template)


@router.post("/{template_id}/reject", response_model=AgentTemplateResponse)
async def reject_template(
    template_id: str,
    reject_data: TemplateRejectRequest,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> AgentTemplateResponse:
    """Reject a pending template with reason (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)

    try:
        template = await template_service.reject_template(template_id, reject_data.reason)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e

    logger.info(f"Template '{template.name}' rejected by {current_user.email}: {reject_data.reason}")

    return AgentTemplateResponse.model_validate(template)


@router.post("/{template_id}/feature", response_model=AgentTemplateResponse)
async def feature_template(
    template_id: str,
    feature_data: TemplateFeatureRequest,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> AgentTemplateResponse:
    """Set featured status for a template (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)

    try:
        template = await template_service.set_featured(template_id, feature_data.featured)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e

    action = "featured" if feature_data.featured else "unfeatured"
    logger.info(f"Template '{template.name}' {action} by {current_user.email}")

    return AgentTemplateResponse.model_validate(template)


@router.patch("/admin/{template_id}", response_model=AgentTemplateResponse)
async def admin_update_template(
    template_id: str,
    template_data: AgentTemplateUpdate,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> AgentTemplateResponse:
    """Update any template (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)

    try:
        template = await template_service.update_template(template_id, template_data)
    except (NotFoundError, ValidationError) as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        ) from e

    return AgentTemplateResponse.model_validate(template)


@router.delete("/admin/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
async def admin_delete_template(
    template_id: str,
    db: Annotated[AsyncSession, Depends(get_tenant_db_dependency)],
    current_user: Annotated[User, Depends(require_super_admin)],
    tenant_id: Annotated[str, Depends(get_tenant_id)],
) -> None:
    """Delete any template (admin only)."""
    TenantContext.set(tenant_id)

    template_service = AgentTemplateService(db)

    try:
        await template_service.delete_template(template_id)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e

    logger.info(f"Template '{template_id}' deleted by {current_user.email}")
